src/Ztm.WebApi/AddressPools/SqlReceivingAddressStorage.cs

```diff
@@ -0,0 +1,202 @@
+
+                return ToDomain(recAddress.Entity);
+            }
+        }
+
+        public async Task<ReceivingAddress> GetAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            {
+                var recv = await db.ReceivingAddresses
+                    .Include(e => e.ReceivingAddressReservations)
+                    .FirstAsync(r => r.Id == id);
+
+                return ToDomain(recv);
+            }
+        }
+
+        public Task<IEnumerable<ReceivingAddress>> ListReceivingAddressAsync(AddressFilter filter, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            {
+                IQueryable<ReceivingAddressModel> query = db.ReceivingAddresses.Include(a => a.ReceivingAddressReservations);
+
+                if (filter.HasFlag(AddressFilter.Available))
+                {
+                    query = query.Where(a => !a.IsLocked);
+                }
+
+                if (filter.HasFlag(AddressFilter.NeverUsed))
+                {
+                    query = query.Where(a => !a.ReceivingAddressReservations.Any());
+                }
+
+                return Task.FromResult<IEnumerable<ReceivingAddress>>(query.Select(r => ToDomain(r)).ToList());
+            }
+        }
+
+        public async Task ReleaseAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            using (var tx = db.Database.BeginTransaction())
+            {
+                var reservation = await db.ReceivingAddressReservations
+                    .Include(r => r.ReceivingAddress)
+                    .FirstAsync(r => r.Id == id);
+
+                if (reservation == null)
+                {
+                    throw new KeyNotFoundException("The reservation id is not found.");
+                }
+
+                if (reservation.ReleasedAt != null)
+                {
+                    throw new InvalidOperationException("The reservation is already released.");
+                }
+
+                reservation.ReleasedAt = DateTime.UtcNow;
+                reservation.ReceivingAddress.IsLocked = false;
+
+                await db.SaveChangesAsync();
+                tx.Commit();
+            }
+        }
+
+        public async Task<ReceivingAddressReservation> TryLockAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.databaseFactory.CreateDbContext())
+            using (var tx = db.Database.BeginTransaction())
+            {
+                var recv = await db.ReceivingAddresses.FirstAsync(a => a.Id == id, cancellationToken);
+                if (recv == null)
+                {
+                    return null;
+                }
+
+                if (recv.IsLocked)
+                {
+                    return null;
```

Comment:
We should throw instead if it is locked so the caller can different between not found and locked.